hoomd/GSDDumpWriter.h

```diff
@@ -153,49 +124,109 @@ class PYBIND11_EXPORT GSDDumpWriter : public Analyzer
+    /// Flags for dynamic/default bitsets.
+    struct gsd_flag
+        {
+        enum Enum
+            {
+            position,
+            orientation,
+            types,
+            type,
+            mass,
+            charge,
+            diameter,
+            body,
+            inertia,
+            velocity,
+            angmom,
+            image,
+            };
+        };
+
+    /// Number of entires in the gsd_flag enum.
+    static const unsigned int n_gsd_flags = 12;
+
+    /// Store a GSD frame for writing.
+    /** Local frames store particles local to the rank, sorted in ascending tag order.
+        Global frames store the entire system, sorted in ascending tag order.
+
+        Entries with 0 sized vectors should not be written to the file. Some ranks may have
+        0 particles while others have N: track which fields are present with
+        `particle_data_present` to enable global communication.
+
+        Note: In the first implementation, only particle data is local/global .
+        The bond/angle/dihedral/etc... data stored in the *local* frame is actually global.
+    */
+    struct GSDFrame
+        {
+        uint64_t timestep;
+        BoxDim global_box;
+
+        std::vector<unsigned int> particle_tags;
+
+        SnapshotParticleData<float> particle_data;
+        BondData::Snapshot bond_data;
+        AngleData::Snapshot angle_data;
+        DihedralData::Snapshot dihedral_data;
+        ImproperData::Snapshot improper_data;
+        ConstraintData::Snapshot constraint_data;
+        PairData::Snapshot pair_data;
+
+        /// Bit flags indicating which particle data fields are present (index by gsd_flag)
+        std::bitset<n_gsd_flags> particle_data_present;
+        };
+
     private:
-    std::string m_fname;            //!< The file name we are writing to
-    std::string m_mode;             //!< The file open mode
-    bool m_truncate = false;        //!< True if we should truncate the file on every analyze()
-    bool m_is_initialized = false;  //!< True if the file is open
-    bool m_write_attribute = false; //!< True if attributes should be written
-    bool m_write_property = false;  //!< True if properties should be written
-    bool m_write_momentum = false;  //!< True if momenta should be written
-    bool m_write_topology = false;  //!< True if topology should be written
-    bool m_write_diameter = false;  //!< True if the diameter attribute should be written
-    gsd_handle m_handle;            //!< Handle to the file
+    std::string m_fname;           //!< The file name we are writing to
+    std::string m_mode;            //!< The file open mode
+    bool m_truncate = false;       //!< True if we should truncate the file on every analyze()
+    bool m_is_initialized = false; //!< True if the file is open
+    bool m_write_topology = false; //!< True if topology should be written
+    bool m_write_diameter = false; //!< True if the diameter attribute should be written
```

Comment:
Because I wish the user to opt-in to writing diameters on both frame 0 and subsequent frames. So far, we only have 1 user who has expressed an interest in this: #1266